test/e2e/helpers.go

```diff
@@ -2,65 +2,166 @@ package e2e
-	// Check if Operator's service monitor service is available
-	_, err := caCtx.Clients.Kube.CoreV1().Services("openshift-serverless").Get(context.Background(), "knative-openshift-metrics2", meta.GetOptions{})
+func getPrometheusRoute(caCtx *test.Context) *v1.Route {
+	r, err := caCtx.Clients.Route.Routes("openshift-monitoring").Get(context.Background(), "prometheus-k8s", meta.GetOptions{})
 	if err != nil {
-		caCtx.T.Fatalf("Error getting service monitor service: %v", err)
+		caCtx.T.Fatalf("Error getting Prometheus route %v", err)
 	}
-	metricsURL, err := url.Parse(metricsPath)
+	return r
+}
+
+func VerifyHealthStatusMetric(caCtx *test.Context, label string, expectedValue string) {
+	pc := NewPrometheusInfo(caCtx)
+	path := fmt.Sprintf("%s%s", pc.queryPath, url.QueryEscape(fmt.Sprintf(`knative_up{type="%s"}`, label)))
+	metricsURL, err := url.Parse(path)
 	if err != nil {
 		caCtx.T.Fatalf("Error parsing url for metrics: %v", err)
 	}
-	expectedStr := fmt.Sprintf(`knative_up{type="%s"} %d`, metricLabel, expectedValue)
 	// Wait until the endpoint is actually working and we get the expected value back
 	_, err = pkgTest.WaitForEndpointState(
 		context.Background(),
 		caCtx.Clients.Kube,
 		caCtx.T.Logf,
 		metricsURL,
-		pkgTest.EventuallyMatchesBody(expectedStr),
+		eventuallyMatchesValue(expectedValue),
 		"WaitForMetricsToServeText",
-		true)
+		true, pc.options...)
+	if err != nil {
+		caCtx.T.Fatalf("Failed to access the Prometheus API endpoint and get the metric value expected: %v", err)
+	}
+}
+
+func VerifyServingControlPlaneMetrics(caCtx *test.Context) {
+	pc := NewPrometheusInfo(caCtx)
+	servingMetrics := []string{
+		"activator_client_results",
+		"autoscaler_actual_pods",
+		"hpaautoscaler_client_latency_bucket",
+		"controller_client_latency_bucket",
+		"domainmapping_client_latency_bucket",
+		"domainmapping_webhook_client_latency_bucket",
+		"webhook_client_latency_bucket",
+	}
+	for _, metric := range servingMetrics {
+		path := fmt.Sprintf("%s%s", pc.queryPath, metric)
+		metricsURL, err := url.Parse(path)
+		if err != nil {
+			caCtx.T.Fatalf("Error parsing url for metrics %v", err)
+		}
+		// Wait until the endpoint is actually working and we get the expected value back
+		_, err = pkgTest.WaitForEndpointState(
+			context.Background(),
+			caCtx.Clients.Kube,
+			caCtx.T.Logf,
+			metricsURL,
+			pkgTest.EventuallyMatchesBody(metric),
+			"WaitForMetricsToServeText",
+			true, pc.options...)
+		if err != nil {
+			caCtx.T.Fatalf("Failed to access the Prometheus API endpoint and get the metric value expected: %v", err)
+		}
+	}
+}
+
+func getBearerTokenForPrometheusAccount(caCtx *test.Context) string {
+	sa, err := caCtx.Clients.Kube.CoreV1().ServiceAccounts("openshift-monitoring").Get(context.Background(), "prometheus-k8s", meta.GetOptions{})
+	if err != nil {
+		caCtx.T.Fatalf("Error getting service account prometheus-k8s: %v", err)
```

Comment:
It feels a little unexpected to me, that these helper functions take the entire test.Context and fail the test immediately. How'd you feel about making them take `kube.Interface` (the client) and returns `(string, error)` and then fail the test in the `VerifyXXX` functions.

Reply:
I thought about this but wanted just to fail as soon as I cannot move forward don think it is unexpected in this context.

Reply:
It was unexpected to me and does hamper the pluggability of the code, but not a strong opinion, it's fine to leave as-is.